prov/util/src/util_wait.c

```diff
@@ -424,3 +428,210 @@ int ofi_wait_fd_open(struct fid_fabric *fabric_fid, struct fi_wait_attr *attr,
 	free(wait);
 	return ret;
 }
+
+static void util_wait_yield_signal(struct util_wait *util_wait)
+{
+	struct util_wait_yield *wait_yield;
+
+	wait_yield = container_of(util_wait, struct util_wait_yield, util_wait);
+
+	fastlock_acquire(&wait_yield->signal_lock);
+	wait_yield->signal = 1;
+	fastlock_release(&wait_yield->signal_lock);
+}
+
+static int util_wait_yield_run(struct fid_wait *wait_fid, int timeout)
+{
+	struct util_wait_yield *wait = container_of(wait_fid,
+			struct util_wait_yield, util_wait.wait_fid);
+	struct ofi_wait_fid_entry *fid_entry;
+	int ret = 0;
+
+	fastlock_acquire(&wait->wait_lock);
+	while (!wait->signal) {
+		dlist_foreach_container(&wait->fid_list,
+					struct ofi_wait_fid_entry,
+					fid_entry, entry) {
+			ret = fid_entry->wait_try(fid_entry->fid);
+			if (ret)
+				return ret;
+		}
```

Comment:
This is the place to yield.  However, we can't do that while holding a spinlock, so something needs to change with the wait_lock here. Wait lock looks like it's serializing access to the fid_list, versus trying to serialize calls to wait.